examples/Reporting/ParallelReportDownload.cs

```diff
@@ -0,0 +1,242 @@
+                        reportId++, responses));
+                }
+            }
+
+            Console.WriteLine($"Awaiting results from {tasks.Count} requests...\n");
+
+            // Proceed only when all requests have completed.
+            await Task.WhenAll(tasks);
+
+            foreach (ReportDownload reportDownload in responses)
+            {
+                // Give a summary report for each successful download.
+                Console.WriteLine(reportDownload);
+            }
+        }
+
+        /// <summary>
+        /// Initiates one asynchronous report download.
+        /// </summary>
+        /// <param name="googleAdsService">The Google Ads service client.</param>
+        /// <param name="customerId">The customer ID from which data is requested.</param>
+        /// <param name="query">The query for the download request.</param>
+        /// <param name="reportId">Integer identifier for this request.</param>
+        /// <param name="responses">Collection of all successful report downloads.</param>
+        /// <returns>The asynchronous operation.</returns>
+        /// <exception cref="GoogleAdsException">Thrown if errors encountered in the execution of
+        ///     the request.</exception>
+        private async Task<SearchStreamStream> DownloadReportAsync(
+            GoogleAdsServiceClient googleAdsService, long customerId, string query, int reportId,
+            ConcurrentBag<ReportDownload> responses)
+        {
+            try
+            {
+                // Issue an asynchronous download request.
+                return await googleAdsService.SearchStreamAsync(
+                    customerId.ToString(), query,
+                    delegate(SearchGoogleAdsStreamResponse resp)
+                    {
+                        // Store the results.
+                        responses.Add(new ReportDownload(customerId, reportId, resp));
+                    }
+                );
+            }
+            catch (AggregateException ae)
+            {
+                Console.WriteLine($"Download failed for report {reportId}!");
+
+                GoogleAdsException gae = GoogleAdsException.FromTaskException(ae);
+
+                if (gae != null)
+                {
+                    Console.WriteLine($"Message: {gae.Message}");
+                    Console.WriteLine($"Failure: {gae.Failure}");
+                    Console.WriteLine($"Request ID: {gae.RequestId}");
+                    throw gae;
+                }
+
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Stores a result from a reporting API call. In this case we simply report a count of
+        /// the responses, but one could also write the response to a database/file, pass the
+        /// response on to another method for further processing, etc.
+        /// </summary>
+        private class ReportDownload
+        {
+            private long CustomerId { get; }
```

Comment:
This is a simple structure to hold some data, right? So 

private class ReportDownload {
  internal long CustomerId { get; set; }
  internal long ReportId { get; set; }
  internal SearchGoogleAdsStreamResponse Response { get; set; }
}

should suffice.

Reply:
Done.

Reply:
Done.